### PR TITLE
Implement memory abstractions

### PR DESCRIPTION
**Pull request by bobakxd**
Implement start version of RAM class which uses preallocated buffer for storing memory.

Size of preallocated memory is contolled using template L parameter which means `limit`.

### CODE REVIEW

File: cpp/memory.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include <cstddef>
+#include <cstdint>
+#include <array>
+#include <vector>
+#include <stdexcept>
+
+template<std::size_t L>
```
Comment (zDonik1) [inline cpp/memory.h]: Please give a proper name to the template argument, such as "Size"

File: cpp/memory.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+template<std::size_t L>
+void RAM<L>::store(std::size_t addr, std::uint16_t v) {
+  check_for_bounds(addr + 1);
+  
+  // Assuming that the value is big-endian.
+  // Firstly extract low byte of the value.
+  std::uint8_t b = v;
+  mem[addr] = b;
```
Comment (zDonik1) [inline cpp/memory.h]: local variable "b" isn't necessary here, and you can use casts instead, as in the following: `static_cast<std::uint8_t>(v)` and `static_cast<std::uint8_t>(v >> 8)`

**Review by zDonik1 — APPROVED** [review]
These changes are slight improvements, otherwise everything is fine :)

File: cpp/memory.h (hunk 1 above)
Comment (bobakxd) [reply cpp/memory.h]: Ohh, so I can name template params like this. I don't know why I had this dumb thought that template arguments should be one letter like `S`, `L` from STL :smile:

File: cpp/memory.h (hunk 2 above)
Comment (bobakxd) [reply cpp/memory.h]: I think implicit conversion should work here. Am I wrong? Maybe there are pitfalls.

**Review by majorkik — APPROVED** [review]

File: cpp/memory.h (hunk 2 above)
Comment (zDonik1) [reply cpp/memory.h]: Yeah, that should work as well, one possible pitfall is that it's not immediately obvious that there is an implicit conversion happening, but that should be minor.

**Pull request merged**